### ISSUE TITLE
TimeoutError: timed out  after thinking for 600 seconds

### ISSUE THREAD

**Issue by watcher60** [aorumbayev/autogpt4all#1]
I managed to get everything installed on a wsl2 Ubuntu 22.04, had to manually install golang and cmake, also unsure why but the script had an access denied when copying the .evnv template file.

While I have validated the LocalAI install with the two POST requests called out on the  of the project of LocalAI and had a response back in  approx a minute, when trying AutoGPT it sets up the goals etc and then sits thinking for 600sec/10mins and then crashes with a timeout.  I have tried installing AutoGPT directly in to windows with LocalAI running the wsl with exactly the same results. 

In the LocalAI window I do see the POST request is received, and when AutoGPT crashes, LocalAI seems to appear to give an empty response (output of the LocalAI at the bottom)

I realize this project is only the install script but was wondering if anyone had any thoughts on how to troubleshoot etc as I'm trying to work out if AutoGPT needs to give LocalAI more time to respond or if it would never respond.

Output from LocalAI (wsl ubuntu)
`:43PM DBG Request received: {"model":"gpt-3.5-turbo","file":"","response_format":"","language":"","prompt":null,"instruction":"","input":null,"stop":null,"messages":[{"role":"system","content":"You are testingAI, an AI designed to write the current date and time to a local text file named datetime.txt\nYour decisions must always be made independently without seeking user assistance. Play to your strengths as an LLM and pursue simple strategies with no legal complications.\n\nGOALS:\n\n1. check the current date and time\n2. write the current date and time to a text file on the local disk\n3. save the text file as datetime.txt\n\nIt takes money to let you run. Your API budget is $20.000\n\nConstraints:\n1. ~4000 word limit for short term memory. Your short term memory is short, so immediately save important information to files.\n2. If you are unsure how you previously did something or want to recall past events, thinking about similar events will help you remember.\n3. No user assistance\n4. Exclusively use the commands listed in double quotes e.g. \"command name\"\n\nCommands:\n1. analyze_code: Analyze Code, args: \"code\": \"\u003cfull_code_string\u003e\"\n2. execute_python_file: Execute Python File, args: \"filename\": \"\u003cfilename\u003e\"\n3. append_to_file: Append to file, args: \"filename\": \"\u003cfilename\u003e\", \"text\": \"\u003ctext\u003e\"\n4. delete_file: Delete file, args: \"filename\": \"\u003cfilename\u003e\"\n5. list_files: List Files in Directory, args: \"directory\": \"\u003cdirectory\u003e\"\n6. read_file: Read file, args: \"filename\": \"\u003cfilename\u003e\"\n7. write_to_file: Write to file, args: \"filename\": \"\u003cfilename\u003e\", \"text\": \"\u003ctext\u003e\"\n8. google: Google Search, args: \"query\": \"\u003cquery\u003e\"\n9. improve_code: Get Improved Code, args: \"suggestions\": \"\u003clist_of_suggestions\u003e\", \"code\": \"\u003cfull_code_string\u003e\"\n10. send_tweet: Send Tweet, args: \"tweet_text\": \"\u003ctweet_text\u003e\"\n11. browse_website: Browse Website, args: \"url\": \"\u003curl\u003e\", \"question\": \"\u003cwhat_you_want_to_find_on_website\u003e\"\n12. write_tests: Write Tests, args: \"code\": \"\u003cfull_code_string\u003e\", \"focus\": \"\u003clist_of_focus_areas\u003e\"\n13. delete_agent: Delete GPT Agent, args: \"key\": \"\u003ckey\u003e\"\n14. get_hyperlinks: Get text summary, args: \"url\": \"\u003curl\u003e\"\n15. get_text_summary: Get text summary, args: \"url\": \"\u003curl\u003e\", \"question\": \"\u003cquestion\u003e\"\n16. list_agents: List GPT Agents, args: () -\u003e str\n17. message_agent: Message GPT Agent, args: \"key\": \"\u003ckey\u003e\", \"message\": \"\u003cmessage\u003e\"\n18. start_agent: Start GPT Agent, args: \"name\": \"\u003cname\u003e\", \"task\": \"\u003cshort_task_desc\u003e\", \"prompt\": \"\u003cprompt\u003e\"\n19. task_complete: Task Complete (Shutdown), args: \"reason\": \"\u003creason\u003e\"\n\nResources:\n1. Internet access for searches and information gathering.\n2. Long Term memory management.\n3. GPT-3.5 powered Agents for delegation of simple tasks.\n4. File output.\n\nPerformance Evaluation:\n1. Continuously review and analyze your actions to ensure you are performing to the best of your abilities.\n2. Constructively self-criticize your big-picture behavior constantly.\n3. Reflect on past decisions and strategies to refine your approach.\n4. Every command has a cost, so be smart and efficient. Aim to complete tasks in the least number of steps.\n5. Write all code to a file.\n\nYou should only respond in JSON format as described below \nResponse Format: \n{\n    \"thoughts\": {\n        \"text\": \"thought\",\n        \"reasoning\": \"reasoning\",\n        \"plan\": \"- short bulleted\\n- list that conveys\\n- long-term plan\",\n        \"criticism\": \"constructive self-criticism\",\n        \"speak\": \"thoughts summary to say to user\"\n    },\n    \"command\": {\n        \"name\": \"command name\",\n        \"args\": {\n            \"arg name\": \"value\"\n        }\n    }\n} \nEnsure the response can be parsed by Python json.loads"},{"role":"system","content":"The current time and date is Tue May 16 17:43:10 2023"},{"role":"system","content":"Your remaining API budget is $20.000\n\n"},{"role":"user","content":"Determine which next command to use, and respond using the format specified above:"}],"stream":false,"echo":false,"top_p":0,"top_k":0,"temperature":0,"max_tokens":2594,"n":0,"batch":0,"f16":false,"ignore_eos":false,"repeat_penalty":0,"n_keep":0,"mirostat_eta":0,"mirostat_tau":0,"mirostat":0,"seed":0}
5:43PM DBG Parameter Config: &{OpenAIRequest:{Model:gpt-3.5-turbo File: ResponseFormat: Language: Prompt:<nil> Instruction: Input:<nil> Stop:<nil> Messages:[] Stream:false Echo:false TopP:0.7 TopK:80 Temperature:0.9 Maxtokens:2594 N:0 Batch:0 F16:false IgnoreEOS:false RepeatPenalty:0 Keep:0 MirostatETA:0 MirostatTAU:0 Mirostat:0 Seed:0} Name: StopWords:[] Cutstrings:[] TrimSpace:[] ContextSize:512 F16:false Threads:4 Debug:true Roles:map[] Embeddings:false Backend: TemplateConfig:{Completion: Chat: Edit:} MirostatETA:0 MirostatTAU:0 Mirostat:0 PromptStrings:[] InputStrings:[] InputToken:[]}
5:43PM DBG Loading models greedly
5:53PM DBG Response: {"object":"chat.completion","model":"gpt-3.5-turbo","choices":[{"message":{"role":"assistant","content":"{ \"thoughts\": { }, \"command\": {} }"}}],"usage":{"prompt_tokens":0,"completion_tokens":0,"total_tokens":0}}
[127.0.0.1]:36418  200  -  POST     /chat/completions`

AutoGPT output (windows)
`Welcome back!  Would you like me to return to being testingAI?
  Asking user via keyboard...
Continue with the last settings?
Name:  testingAI
Role:  an AI designed to write the current date and time to a local text file named datetime.txt
Goals: ['check the current date and time', 'write the current date and time to a text file on the local disk', 'save the text file as datetime.txt']
API Budget: $20.0
Continue (y/n): y
testingAI  has been created with the following details:
Name:  testingAI
Role:  an AI designed to write the current date and time to a local text file named datetime.txt
Goals:
-  check the current date and time
-  write the current date and time to a text file on the local disk
-  save the text file as datetime.txt
Using memory of type:  LocalCache
Using Browser:  chrome
Traceback (most recent call last):
  File "C:\Python311\Lib\site-packages\urllib3\connectionpool.py", line 449, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "C:\Python311\Lib\site-packages\urllib3\connectionpool.py", line 444, in _make_request
    httplib_response = conn.getresponse()
                       ^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\http\client.py", line 1375, in getresponse
    response.begin()
  File "C:\Python311\Lib\http\client.py", line 318, in begin
    version, status, reason = self._read_status()
                              ^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\http\client.py", line 279, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\socket.py", line 706, in readinto
    return self._sock.recv_into(b)
           ^^^^^^^^^^^^^^^^^^^^^^^
TimeoutError: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python311\Lib\site-packages\requests\adapters.py", line 486, in send
    resp = conn.urlopen(
           ^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\urllib3\connectionpool.py", line 787, in urlopen
    retries = retries.increment(
              ^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\urllib3\util\retry.py", line 550, in increment
    raise six.reraise(type(error), error, _stacktrace)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\urllib3\packages\six.py", line 770, in reraise
    raise value
  File "C:\Python311\Lib\site-packages\urllib3\connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
                       ^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\urllib3\connectionpool.py", line 451, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "C:\Python311\Lib\site-packages\urllib3\connectionpool.py", line 340, in _raise_timeout
    raise ReadTimeoutError(
urllib3.exceptions.ReadTimeoutError: HTTPConnectionPool(host='localhost', port=8080): Read timed out. (read timeout=600)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python311\Lib\site-packages\openai\api_requestor.py", line 516, in request_raw
    result = _thread_context.session.request(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\requests\sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\requests\sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\requests\adapters.py", line 532, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPConnectionPool(host='localhost', port=8080): Read timed out. (read timeout=600)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "c:\Users\a9dqmzz\Auto-GPT\autogpt\__main__.py", line 5, in <module>
    autogpt.cli.main()
  File "C:\Python311\Lib\site-packages\click\core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\click\core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\click\core.py", line 1635, in invoke
    rv = super().invoke(ctx)
         ^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\click\core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\click\core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\click\decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\a9dqmzz\Auto-GPT\autogpt\cli.py", line 90, in main
    run_auto_gpt(
  File "c:\Users\a9dqmzz\Auto-GPT\autogpt\main.py", line 186, in run_auto_gpt
    agent.start_interaction_loop()
  File "c:\Users\a9dqmzz\Auto-GPT\autogpt\agent\agent.py", line 113, in start_interaction_loop
    assistant_reply = chat_with_ai(
                      ^^^^^^^^^^^^^
  File "c:\Users\a9dqmzz\Auto-GPT\autogpt\llm\chat.py", line 244, in chat_with_ai
    assistant_reply = create_chat_completion(
                      ^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\a9dqmzz\Auto-GPT\autogpt\llm\llm_utils.py", line 166, in create_chat_completion
    response = api_manager.create_chat_completion(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\a9dqmzz\Auto-GPT\autogpt\llm\api_manager.py", line 55, in create_chat_completion
    response = openai.ChatCompletion.create(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\openai\api_resources\chat_completion.py", line 25, in create
    return super().create(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\openai\api_resources\abstract\engine_api_resource.py", line 153, in create
    response, _, api_key = requestor.request(
                           ^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\openai\api_requestor.py", line 216, in request
    result = self.request_raw(
             ^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\openai\api_requestor.py", line 526, in request_raw
    raise error.Timeout("Request timed out: {}".format(e)) from e
openai.error.Timeout: Request timed out: HTTPConnectionPool(host='localhost', port=8080): Read timed out. (read timeout=600)
Press any key to continue . . .
`

**Comment by aorumbayev** [comment]
@watcher60 indeed the repo is just the bare minimum wip setup to make localai + gpt4all + autogpt to work together. I suppose you can try playing with slightly smaller models, i used vicuna 7b on m2 and it was around 30 seconds per response on m2 and worked mostly ok but few times was failing on auto agent definition stage. Alternatively a longer route is tweaking the autogpt code and increase the default timeouts past 600 secs.

**Comment by watcher60** [comment]
@aorumbayev  Thank you for the tip on using a smaller model. I did manage to track down where the timeout value is coming from. Auto-GPT is using the openAI package, specifically api_requestor.py, which has the timeout value set near the top. As a quick test I edited this in my python site packages to 6000 rather than 600 and on my test run its now been "thinking" for past the magic 600 seconds.

Just to advise this changed worked (oddly I seem to have better response times with the larger model). Obviously editing a site package is not best practice so will have to see how to see if there is another option long term, though with llama now supporting GPU this may be a non-issue for most long term)
